README.org
#+title: Swinfo - Get software information
#+author: iseeu
#+date: 2021-06-27 00:08:16 +0300

* Introduction

Swinfo (software information) is an Emacs package that returns
information for software in different methods, like sending the info to
the clipboard, echo or insert it in the current buffer. This package can
be helpful in situation where you want to report an issue but the
information for the tool in question is needed to help fix or narrow it
down.

There are five categories to get information from:

1. Static:

   This is for software that the chance they get updated are low or the
   user choose not to do so. Sometimes it better to feed the information
   manually than go with the hard or maybe tricky road to get the
   information from a specific software.

2. Repository:

   When you are using a software by compiling it from its source Git
   repository, it's easy to know which latest commit hash and date for a
   single repository; but what if you are having many repositories you
   want to check their information at once, it will take time to do so.
   This package can help you in this regards; you only need to add minor
   information about these repositories to get the information you are
   after.

3. Built-in package:

   You can provide a list of built-in packages and shape how their
   information is presented. Currently, the support for this category is
   shallow.

4. External package:

   This is for external packages installed by =package.el= library which
   are exist in =elpa= directory.

5. Unix tool:

   Unix tools are a lot and they have different options to output their
   version info. So, this package is doing a smart check and return the
   info as to let you not thinking much about the method but how to
   present the info.

* Installation

This package is currently not available in Melpa or any other package
archives. It's too young to put it in popular archives as things might
change frequently.

To use this package you need to clone it and then require it in your
Emacs init file.

1. Clone the package repository to a location that =load-path= variable
   is looking into:

   #+begin_src shell
git clone https://github.com/iSeeU816/swinfo
   #+end_src

2. There are two ways to setup this package:
   1. Emacs vanilla way:

      #+begin_src emacs-lisp
(add-to-list 'load-path
             "/path/where/you/cloned/the/repository/swinfo")
(require 'swinfo)
(with-eval-after-load 'swinfo
  (setq swinfo-static-list
        '((foo . (:sw-name "Foo" :sw-ver "0.8.16"))
          (bar . (:sw-name "Bar" :sw-ver "0.0.1"))))
  ;; Add other list customization here.

  (let ((map global-map))
    (define-key map (kbd "C-c i s") #'swinfo)
    (define-key map (kbd "C-c i i") #'swinfo-insert-info)
    (define-key map (kbd "C-c i e") #'swinfo-echo-info)
    (define-key map (kbd "C-c i k") #'swinfo-kill-info)))
      #+end_src

   2. [[https://github.com/jwiegley/use-package][use-package]] way:

      #+begin_src emacs-lisp
(use-package swinfo
  :load-path "/path/where/you/cloned/the/repository/swinfo"
  :config
  (setq swinfo-static-list
        '((foo . (:sw-name "Foo" :sw-ver "0.8.16"))
          (bar . (:sw-name "Bar" :sw-ver "0.0.1"))))
  :bind (("C-c i s" . swinfo)
         ("C-c i i" . swinfo-insert-info)
         ("C-c i e" . swinfo-echo-info)
         ("C-c i k" . swinfo-kill-info)))
      #+end_src

Note that these are examples setup, you should use real world data.
Refer to usage section below to know more about how to use this package.

* Usage

For minimal usage of this package, you can require its library with
=(require 'swinfo)= and only call =M-x swinfo= to choose from installed
external packages or even type a Unix tool name to insert their
information.

Here we go with a detailed instructions with some examples on how to use
this package:

** Static

1. First you set value for =swinfo-static-list= variable and evaluate
   it. You can add this to your Emacs init file:

   #+begin_src emacs-lisp
(setq swinfo-static-list
      '((foo . (:sw-name "Foo" :sw-ver "0.8.16"))
        (bar . (:sw-name "Bar" :sw-ver "0.0.1"))))
   #+end_src

2. Call =M-x swinfo= and type =foo= or select it from the list. You
   should have something like this:
   - =foo: Foo 0.8.16=.
3. You can use the Emacs Lisp form =(swinfo '(foo))= and evaluate it to
   get the same result as above. To evaluate it, just put the point
   after the closing parenthesis and hit =C-x C-e= or call =M-x
   eval-last-sexp= command.

** Repository

This shares the same structure as static category, except this has an
extra property key named =:dir=. That property key is essential to know
where the repository is located so Swinfo knows what =default-directory=
should be used when calling shell command related to Git.

1. Set =swinfo-repo-list= variable like this and evaluate it as
   mentioned in static section above:

   #+begin_src emacs-lisp
(setq swinfo-repo-list
      '((emacs-config . (:dir "~/.emacs.d"))
        (emacs-src . ( :dir "~/my_clone/emacs-src"
                       :sw-name "Emacs"
                       :command (funcall (lambda () (format "%s" emacs-version)))))))
   #+end_src

2. When calling =M-x swinfo= and then typing =emacs-config,emacs-src=
   and then hitting =RET=, the output should be like this:

   #+begin_src text
emacs-config: rev eddcda0520 on 2021-06-14
emacs-src: Emacs 28.0.50; rev cee502ea18 on 2021-05-29
   #+end_src

*Notes:*

- Notice how only =:dir= property key is defined in =emacs-config= cons
  list; which means no extra info should be printed. This type of style
  work well for repositories that are not related to a software.
- Not like the other example, where commit hash and date info are not
  enough; so the need to print the software name and version sometimes
  necessary and add such context about the repository in question.
- The comma character between software names in step 2 above is the
  default value of =crm-separator= variable. This is valid when dealing
  with a command that support multiple inputs, you know this if the
  command uses =completing-read-multiple= function instead of
  =completing-read= function.

** Built-in package

A simple category to shape the way that some built-in packages return
their versions.

1. Set =swinfo-built-in-package-list= variable to something like this
   and evaluate it:

   #+begin_src emacs-lisp
(setq swinfo-built-in-package-list
      '((org . (:sw-name "Org" :command (funcall org-version)))))
   #+end_src

2. Call =M-x swinfo= and type =org= to get info like this:
   - =org: Org 9.4.4=.

** External packages

This category currently only concern about packages installed by
=package.el= library which they installed in =elpa= directory.

No needed setup for this category, just call =M-x swinfo= and type a
package name to get its info. For example =magit= which the output
should look like this:

#+begin_src text
magit: magit-20210531.1524; rev 4a8646a30d
#+end_src

Or you can use an Emacs Lisp form =(swinfo '(magit))= to get the same
result.

** Unix tool

This is a powerful category to have. It simply returns the first line of
the output that Unix tool version command sends out.

Call =M-x swinfo= and then type a Unix tool to get its version info. For
example =git= should returns =git: git version 2.32.0=.

When using Emacs Lisp form, you can double quote the Unix tool name, so
the above example should look like this =(swinfo '("git"))= and the
output is the same.

*IMPORTANT NOTE:* The name of Unix tool that you provide must match the
 full name. No fuzzy search is applied here, so =(swinfo '("gi"))=
 signals an error if you meant to get Git version and there is no =gi=
 tool is installed. To check of what you should type, run this command
 in a shell ~$ type git~ which should NOT signal an error but output the
 path for the tool; of course, change =git= here with the Unix tool in
 mind.

** Mix

You can mix categories together in one command to get the information
you're after.

For example =(swinfo '(emacs-src "git" magit))= should output this:

#+begin_src text
emacs-src: Emacs 28.0.50; rev cee502ea18 on 2021-05-29
git: git version 2.32.0
magit: magit-20210531.1524; rev 4a8646a30d
#+end_src

The order of the output is the same order in the command.

** Output methods

Swinfo provides three methods to get the information.

*** Kill (send info to the clipboard)

You can send the info to the clipboard by doing any of the following
items:

+ Hit =C-u= (control-u) and then call Swinfo command by either =M-x
  swinfo= or invoke the key binding that you might assigned it to
  =swinfo= command.

+ In Emacs Lisp form: =(swinfo '(magit) 'kill)=. Notice the =kill=
  symbol.

*** Echo (echo info to echo area)

You might wants to check a version info for a package, this is what
echo feature is for.

+ Hit =C-u C-u= (control-u twice) and then call Swinfo command by either
  =M-x swinfo= or invoke the key binding that you might assigned it to
  =swinfo= command.

+ Same as Kill method, you just need to use =echo= symbol instead.
  =(swinfo '(magit) 'echo)=.

*** Insert (insert info in the current buffer)

You're in middle of reporting something and wants to mention what
version of software you're using as it might adds context to the issue
in mind; here where this method shines.

+ With this method, you just call =M-x swinfo= or invoke the key binding
  to =swinfo= command.

+ You can use four different forms of Emacs Lisp with this method:

  1. =(swinfo '(magit) 'insert)=: Use =insert= symbol for the optional
     argument.
  2. =(swinfo '(magit) t)=.
  3. =(swinfo '(magit) nil)=.
  4. =(swinfo '(magit))=: You can omit the optional argument all
     together.

* Contribute

Please report a bug if you encounter one or request a feature if it's in
this project scope.

Of course, you can contribute to the code if you want to. Thanks!

* License

GNU General Public License, version 3.0.


CHANGELOG.org
#+title: Swinfo change log
#+author: iseeu
#+date: 2021-07-01 06:51:44 +0300

* Change log

** Unreleased

- Fix an error when trying to get a Unix tool info.
- Rename swinfo-yank-info to swinfo-kill-info.
  - And update readme accordingly.

** Version 0.0.1 (2021-07-01)

- Initial public version.


swinfo.el
;;; swinfo.el --- get software information -*- lexical-binding: t; -*-

;; Copyright (C) 2021  iSeeU

;; Author: iSeeU
;; Created: 2021-06-03 07:12:19 +0300
;; Version: 0.0.1
;; Keywords: software info information version
;; Homepage: https://github.com/iSeeU816/swinfo

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Swinfo (software information) is an Emacs package that returns
;; information for software in different methods, like sending the info
;; to the clipboard, echo or insert it in the current buffer. This
;; package can be helpful in situation where you want to report an issue
;; but the information for the tool in question is needed to help fix or
;; narrow it down.

;; See this package's README.org file for more info.

;;; Code:

(require 'package)

(defconst swinfo-version "0.0.1"
  "The version of Swinfo.")

;;;; Options

(defvar swinfo-static-list '()
  "List of software that are mostly in static phase.

An association list (Alist) where CAR is a symbol label for a
software, and CDR is a property list (Plist) that have (KEY1
VALUE1 KEY2 VALUE2 ...) style.

Currently, only `:sw-name' and `:sw-ver' property keys are
supported. The values of them should be string.

This is mostly for software that met these criteria:

- The software won't be updated anymore or the user choose not to
  do so.

- It's hard to parse the information or it's not worth it; so
  simply doing it manually is much faster.

- No automatic way to parse the information.")

(defvar swinfo-repo-list '()
  "List of software repositories to parse info from them.

An association list (Alist) where CAR is a symbol label for a
repository, and CAR is a property list (Plist) that have (KEY1
VALUE1 KEY2 VALUE2 ...) style.

Currently, only `:dir', `:sw-name' and `:command' property keys are
supported.

`:dir': A string of a repository absolute path.

`:sw-name' (optional): A string of software name.

`:command' (optional): Call a command with `funcall' function
which must return a string. It's useful to parse information that
can't be parsed from the repository.

Note that this can be used for repositories that are not related
to any software. It's a way of getting latest commit hash and
date for them.")

(defvar swinfo-built-in-package-list '()
  "List of built-in packages.

An association list (Alist) where CAR is a symbol label for a
repository, and CAR is a property list (Plist) that have (KEY1
VALUE1 KEY2 VALUE2 ...) style.

Currently, only `:sw-name' and `:command' property keys are
supported.

`:sw-name': A string of software name.

`:command': Call a command with `funcall' function which must
return a string.")

;;;; Helpers

(defun swinfo--call-process (command &rest args)
  "Return output (as string) of shell COMMAND that can be called
with optional arguments ARGS. Consult `call-process' for more
info."
  (with-temp-buffer
    (list (apply 'call-process command nil '(t nil) nil args)
          (buffer-string))))

(defun swinfo--get-first-line (str)
  "Return first line of multiple lines string STR.

This is useful to only get the first line of a shell command
output."
  (replace-regexp-in-string "\n.*" "" str))

(defun swinfo--plist-get (alist-list alist-key plist-prop)
  "Get value of PLIST-PROP key that is nested under ALIST-KEY for ALIST-LIST.

Suppose you have something like this:

  (setq swinfo-repo-list
        '((repo-name . (:dir \"/tmp/foo/repo-name\" :sw-name \"Foo\"))))

And you want to get \"Foo\", so you call:

  (swinfo--plist-get swinfo-repo-list 'repo-name :sw-name)"
  (plist-get (alist-get alist-key alist-list nil nil 'equal) plist-prop))

;;;; Info

;;;;; Static

(defun swinfo-static-info (name)
  "Return information about static software that goes with the NAME.

Currently, only software name and version are supported."
  (when swinfo-static-list
    (let ((sw-name (swinfo--plist-get swinfo-static-list name :sw-name))
          (sw-ver (swinfo--plist-get swinfo-static-list name :sw-ver)))
      (format "%s:%s%s" name
              (if sw-name (concat " " sw-name) "")
              (if sw-ver (concat " " sw-ver) "")))))

;;;;; Repository

(defun swinfo-repo-commit-hash (repo-name)
  "Return repository REPO-NAME commit hash (10 digits)."
  (with-temp-buffer
    (let* ((default-directory (swinfo--plist-get swinfo-repo-list repo-name :dir))
           (latest-commit-hash
            (cadr (swinfo--call-process "git" "rev-parse" "HEAD")))
           (latest-commit-hash-short (substring latest-commit-hash 0 10)))
      (format "%s" latest-commit-hash-short))))

(defun swinfo-repo-commit-date (repo-name)
  "Return repository REPO-NAME commit date (ISO format)."
  (with-temp-buffer
    (let* ((default-directory (swinfo--plist-get swinfo-repo-list repo-name :dir))
           (latest-commit-date
            ;; Using `string-trim' to get rid of the newline at the end
            ;; of result string.
            (string-trim
             (cadr (swinfo--call-process "git" "log" "-1"
                                         "--date=short" "--format=%cd")))))
      (format "%s" latest-commit-date))))

(defun swinfo-repo-info (repo-name)
  "Return information about repository REPO-NAME.

Supported information so far:

- Repository name (as the user label it in `swinfo-repo-list'
  variable).

- Software name (if any provided).

- A command output, something like software version that related
  to the repository (if any).

- Latest commit hash and date."
  (let* ((commit-hash (swinfo-repo-commit-hash repo-name))
         (commit-date (swinfo-repo-commit-date repo-name))
         (sw-name (swinfo--plist-get swinfo-repo-list repo-name :sw-name))
         (command (swinfo--plist-get swinfo-repo-list repo-name :command))
         (command-result (when command (apply command))))
    (if command-result
        (format "%s: %s %s; rev %s on %s"
                repo-name sw-name command-result
                commit-hash commit-date)
      (format "%s: rev %s on %s" repo-name commit-hash commit-date))))

;;;;; Built-in package

(defun swinfo-built-in-package-info (pkg-name)
  "Return information about built-in package PKG-NAME.

Supported information so far:

- Software name.

- A command output, something like software version."
  (let* ((sw-name (swinfo--plist-get
                   swinfo-built-in-package-list pkg-name :sw-name))
         (command (swinfo--plist-get
                   swinfo-built-in-package-list pkg-name :command))
         (command-result (when command (apply command))))
    (format "%s: %s %s" pkg-name sw-name command-result)))

;;;;; External package

(defun swinfo--package-desc (pkg-name)
  "Return package name as symbol for PKG-NAME from `package-alist'
variable."
  ;; Using `package--alist' function is to make sure that
  ;; `pacakge-alist' variable value is not `nil'.
  (car (cdr (assq pkg-name (package--alist)))))

(defun swinfo--package-desc-extras (pkg-name slot)
  "Return value of SLOT for PKG-NAME from `package-alist' variable
with the help of `package-desc-extras' function."
  (cdr (assoc slot (package-desc-extras
                    (car (cdr (assq pkg-name (package--alist))))))))

(defun swinfo-package-info (pkg-name)
  "Return information about package PKG-NAME.

Supported information so far:

- Package name to be used as label.

- Full name that is package name and its version as latest's
  commit date. Something like `foo-20210624.1859'.

- Latest commit hash (10 digits)."
  (when (memq pkg-name (mapcar #'car (package--alist)))
    (let* ((name (package-desc-name (swinfo--package-desc pkg-name)))
           (full-name (package-desc-full-name (swinfo--package-desc pkg-name)))
           (commit-hash (swinfo--package-desc-extras pkg-name :commit))
           (commit-hash-short (substring commit-hash 0 10)))
      (format "%s: %s; rev %s"
              pkg-name full-name commit-hash-short))))

;;;;; Unix tool

(defun swinfo-unix-tool-info (name)
  "Return Unix tool NAME information.

Currently, it only returns the first line of what Unix tool
version command output is."
  (let* ((name (if (symbolp name)
                   (symbol-name name)
                 name))
         (output
          (cond
           ((eq 0 (car (swinfo--call-process name "--version")))
            (swinfo--get-first-line
             (cadr (swinfo--call-process name "--version"))))
           ((eq 0 (car (swinfo--call-process name "-v")))
            (swinfo--get-first-line
             (cadr (swinfo--call-process name "-v"))))
           ((eq 0 (car (swinfo--call-process name "-V")))
            (swinfo--get-first-line
             (cadr (swinfo--call-process name "-V")))))))
    (format "%s: %s" name output)))

;;;; Misc

(defun swinfo-version (&optional full)
  "Echo Swinfo version information.

When FULL argument is non-nil print Swinfo with the version."
  (when swinfo-version
    (if full
        (format "Swinfo %s" swinfo-version)
      swinfo-version)))

;;;; Output

(defvar swinfo-software-list '()
  "A list to gather software names in as symbols.")

(defun swinfo--combine-list ()
  "Combine lists to have one list of software names as symbols.

Note that only installed packages (`package-activated-list') are
gathered and not all packages as what `package-alist' variable is
for."
  (setq swinfo-software-list
        (append (mapcar #'car swinfo-static-list)
                (mapcar #'car swinfo-repo-list)
                package-activated-list)))

(defun swinfo--info (name)
  "Loop through NAME (list of software names) to gather information
about them and set the result to `swinfo-info' variable."
  (let ((name (reverse name))
        item
        info)
    (while name
      (setq item (pop name))
      (cond
       ((equal item (car (assoc item swinfo-static-list)))
        (push (swinfo-static-info item) info))
       ((equal item (car (assoc item swinfo-repo-list)))
        (push (swinfo-repo-info item) info))
       ((equal item (car (assoc item swinfo-built-in-package-list)))
        (push (swinfo-built-in-package-info item) info))
       ((memq item package-activated-list)
        (push (swinfo-package-info item) info))
       ((eq 0 (shell-command (format "type %s" item)))
        ;; To hide the message in echo area that resulted from
        ;; `shell-command' function above.
        (message "")
        (push (swinfo-unix-tool-info item) info))
       (t (message "swinfo: Nothing matches `%s'." item))))
    (setq swinfo-info (string-join info "\n"))))

(defun swinfo--get-info (get-method)
  "Get software information in different methods.

If called with one \\[universal-argument] or GET-METHOD argument
value is `kill' (symbol) then send info to `kill-ring', but if
called with two \\[universal-argument] or GET-METHOD argument
value is `echo' (symbol) then echo info to echo area. Otherwise,
insert info in current buffer. The latter case is when GET-METHOD
argument value is `nil', `t' or `insert' (symbol).

This function is meant to be used in `swinfo' function."
  (cond
   ((or (equal current-prefix-arg '(4))
        (eq get-method 'kill))
    (kill-new swinfo-info)
    (message "swinfo: Killed (copied) software info into the kill ring."))
   ((or (equal current-prefix-arg '(16))
        (eq get-method 'echo))
    (let ((single-line
           (replace-regexp-in-string "\n" " -- " swinfo-info)))
      (message "swinfo: %s" single-line)))
   (t (insert swinfo-info))))

(defvar swinfo--software-name-history nil
  "Variable to hold Swinfo history when type/select software names
in minibuffer.")

(defun swinfo-kill-info ()
  "Send software info to `kill-ring'."
  (interactive)
  (let ((current-prefix-arg '(4)))
    (call-interactively #'swinfo)))

(defun swinfo-echo-info ()
  "Echo software info to echo area."
  (interactive)
  (let ((current-prefix-arg '(16)))
    (call-interactively #'swinfo)))

(defun swinfo-insert-info ()
  "Insert software info in the current buffer."
  (interactive)
  (let ((current-prefix-arg nil))
    (call-interactively #'swinfo)))

(defun swinfo (name &optional get-method)
  "Get software information for NAME.

NAME argument is a list. This can contain symbols and strings.
The string part is for Unix tool category.

The optional GET-METHOD argument can be a symbol that represent
the output method, which are `kill', `echo' and `insert' symbols.
For interactive usage, \\[universal-argument] is the one that
represents which method should be used, see `swinfo--get-info'
function for more info about this. If this argument omitted then
`insert' method what will be used.

When called interactively, you can choose multiple candidates by
separate them with `crm-separator' character. For Unix tool, you
can type them as is, as `swinfo--info' function will check them
if no match was found in other categories' list."
  (interactive
   (list
    (progn
      (swinfo--combine-list)
      (let ((name (completing-read-multiple "Software name: " swinfo-software-list
                                            nil nil nil 'swinfo--software-name-history)))
        (and (> (length name) 0) (mapcar #'intern name))))
    current-prefix-arg))
  (if (< (length name) 1)
      (user-error "swinfo: No software name specified")
    (swinfo--combine-list)
    (swinfo--info name)
    (swinfo--get-info get-method)))

;;;; Closing marks

(provide 'swinfo)

;;; swinfo.el ends here


swver.el
;;; swver.el --- Get software version -*- lexical-binding: t; -*-

;; Copyright (C) 2021  iSeeU

;; Author: iSeeU
;; Created: 2021-06-03 07:12:19 +0300
;; Version: 0.0.1a17
;; Keywords: software version

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Get a software version to insert it in a buffer, echo it or even send
;; it to the kill ring.

;;; Code:

(require 'package)

(defconst swver-version "0.0.1a17"
  "The version of Swver.")

(defvar swver-static '()
  "WIP; 2021-06-15 13:50:52 +0300.")

(defvar swver-repo-dir
  '((emacs-config . (dir "~/.emacs.d"))
    (emacs-src . ( dir "~/my_clone/emacs-src"
                   sw-name "Emacs"
                   command (funcall (lambda () (format "%s" emacs-version))))))
  "WIP; 2021-06-04 09:02:26 +0300.")

(defvar swver-built-in-package '()
  "WIP; 2021-06-15 15:26:38 +0300.")

(defun swver--call-process (command &rest args)
  "WIP; 2021-06-14 14:05:41 +0300."
  (with-temp-buffer
    (list (apply 'call-process command nil '(t nil) nil args)
          (buffer-string))))

(defun swver--get-first-line (str)
  "WIP; 2021-06-15 10:42:55 +0300."
  (replace-regexp-in-string "\n.*" "" str))

(defun swver--plist-get (alist-list alist-key plist-prop)
  "WIP; 2021-06-19 15:39:52 +0300."
  (plist-get (alist-get alist-key alist-list nil nil 'equal) plist-prop))

(defun swver--plist-get-prop-p (list name prop)
  "WIP; 2021-06-21 16:34:53 +0300."
  (if (swver--plist-get list name prop) t nil))

(defun swver-static-info (name)
  "WIP; 2021-06-21 15:58:49 +0300."
  (when swver-static
    (let ((sw-name (swver--plist-get swver-static name 'sw-name))
          (sw-ver (swver--plist-get swver-static name 'sw-ver)))
      (format "%s:%s%s" name
              (if sw-name (concat " " sw-name) "")
              (if sw-ver (concat " " sw-ver) "")))))

(defun swver-repo-commit-hash (repo-name)
  "WIP; 2021-06-05 14:14:05 +0300"
  (with-temp-buffer
    (let* ((default-directory (swver--plist-get swver-repo-dir repo-name 'dir))
           (latest-commit-hash
            (cadr (swver--call-process "git" "rev-parse" "HEAD")))
           (latest-commit-hash-short (substring latest-commit-hash 0 10)))
      (format "%s" latest-commit-hash-short))))

(defun swver-repo-commit-date (repo-name)
  "WIP; 2021-06-05 15:04:32 +0300."
  (with-temp-buffer
    (let* ((default-directory (swver--plist-get swver-repo-dir repo-name 'dir))
           (latest-commit-date
            ;; Using `string-trim' to get rid of the newline at the end
            ;; of result string.
            (string-trim
             (cadr (swver--call-process "git" "log" "-1"
                                        "--date=short" "--format=%cd")))))
      (format "%s" latest-commit-date))))

(defun swver-repo-info (repo-name)
  "WIP; 2021-06-05 15:23:07 +0300."
  (let* ((commit-hash (swver-repo-commit-hash repo-name))
         (commit-date (swver-repo-commit-date repo-name))
         (sw-name (swver--plist-get swver-repo-dir repo-name 'sw-name))
         (command (swver--plist-get swver-repo-dir repo-name 'command))
         (command-result (when command (apply command))))
    (if command-result
        (format "%s: %s %s; rev %s on %s"
                repo-name sw-name command-result
                commit-hash commit-date)
      (format "%s: rev %s on %s" repo-name commit-hash commit-date))))

(defun swver--package-desc (pkg-name)
  "WIP; 2021-06-20 18:37:45 +0300."
  (car (cdr (assq pkg-name package-alist))))

(defun swver--package-desc-extras (pkg-name slot)
  "WIP; 2021-06-20 18:57:46 +0300."
  (cdr (assoc slot (package-desc-extras
                    (car (cdr (assq pkg-name package-alist)))))))

(defun swver-package-info (pkg-name)
  "WIP; 2021-06-12 11:14:30 +0300."
  (when (memq pkg-name (mapcar #'car package-alist))
    (let* ((name (package-desc-name (swver--package-desc pkg-name)))
           (full-name (package-desc-full-name (swver--package-desc pkg-name)))
           (commit-hash (swver--package-desc-extras pkg-name :commit))
           (commit-hash-short (substring commit-hash 0 10)))
      (format "%s: %s (rev %s)"
              pkg-name full-name commit-hash-short))))

(defun swver-unix-tool-info (name)
  "WIP; 2021-06-14 17:37:54 +0300."
  (let ((output
         (cond
          ((eq 0 (car (swver--call-process name "--version")))
           (swver--get-first-line
            (cadr (swver--call-process name "--version"))))
          ((eq 0 (car (swver--call-process name "-v")))
           (swver--get-first-line
            (cadr (swver--call-process name "-v"))))
          ((eq 0 (car (swver--call-process name "-V")))
           (swver--get-first-line
            (cadr (swver--call-process name "-V"))))
          (t (message "swver: All conditions failed.")))))
    (format "%s" output)))

(defvar swver-software-list '()
  "WIP; 2021-06-13 11:18:02 +0300.")

(defun swver--combine-list ()
  "WIP; 2021-06-13 11:18:57 +0300."
  (setq swver-software-list
        (append (mapcar #'car swver-static)
                (mapcar #'car swver-repo-dir)
                package-activated-list)))

(defun swver--info (name)
  "WIP; 2021-06-12 13:04:17 +0300."
  (let ((name (reverse name))
        item
        info)
    (message "swver: `%s' type is %s." name (type-of name))
    (while name
      (setq item (pop name))
      (message "swver: `%s' type is %s." item (type-of item))
      (cond
       ((equal item (car (assoc item swver-static)))
        (message "swver: `%s' is a static and its type is %s." item (type-of item))
        (push (swver-static-info item) info))
       ((equal item (car (assoc item swver-repo-dir)))
        (message "swver: `%s' is a repo and its type is %s." item (type-of item))
        (push (swver-repo-info item) info))
       ((equal item (car (assoc item swver-built-in-package)))
        (message "swver: `%s' is a built-in and its type is %s." item (type-of item))
        (push (funcall (cdr (assoc item swver-built-in-package))) info))
       ((memq item package-activated-list)
        (message "swver: `%s' is a package and its type is %s." item (type-of item))
        (push (swver-package-info item) info))
       ((eq 0 (shell-command (format "type %s" item)))
        (push (swver-unix-tool-info item) info))
       (t (message "swver: Nothing matches `%s'." item))))
    (message "swver: `%s'; the type is %s." info (type-of info))
    (setq swver-info (string-join info "\n"))))

(defun swver-get-info (&optional arg)
  "WIP; 2021-06-12 07:47:12 +0300."
  (let ((arg (car current-prefix-arg)))
    (cond
     ((eq arg 4)
      (kill-new swver-info)
      (message "swver: Yanked software info into the kill ring."))
     ((eq arg 16)
      (let ((single-line
             (replace-regexp-in-string "\n" " -- " swver-info)))
        (message "swver: %s" single-line)))
     (t (insert swver-info)))))

(defvar swver--software-name-history nil
  "WIP; 2021-06-13 17:36:55 +0300.")

(defun swver (&rest name)
  "WIP; 2021-06-04 13:29:10 +0300."
  (interactive
   (progn
     (swver--combine-list)
     (let ((name (completing-read-multiple "Software name: " swver-software-list
                                           nil nil nil 'swver--software-name-history)))
       (and (> (length name) 0) (mapcar #'intern name)))))
  (if (< (length name) 1)
      (user-error "swver: No software name specified")
    (message "swver: `%s' and its type is %s" name (type-of name))
    (message "swver: `%s' and its type is %s" (car name) (type-of (car name)))
    (swver--info name)
    (swver-get-info)))

(provide 'swver)

;;; swver.el ends here
